### DISCUSSION TITLE
How to split a form up into multiple files?

### DISCUSSION THREAD

**Discussion by thely** [vueform/vueform#255] [Questions]
I'm loving this library so far - I've had to make similar (and far simpler) tools for myself like it in the past, but they were never this robust. Adam, you're great.

Here's the problem: along with having a lot of fields and two-way data syncing, I've got a lot of function-based conditions, SelectElement conditional lists, reactive component-based content, etc. At about 1300 lines, it's at the point where I'm starting to lose my place in the file. In a different project, I would have broken this up into separate files a long time ago, but I'm still figuring out how to do that with Vueform. While I did start attempting to [follow the instructions on creating a custom Group/ObjectElement](https://vueform.com/recipes/creating-a-custom-element-containing-multiple-elements), I ran into some questions:

1. My form right now is schema-based, and if making custom GroupElements is the easiest way to go about splitting up the form, I'd like to keep using a schema rather than messing with the template. However, some of these custom elements would need certain values passed in as props as well, and you can't have props rely on other props in a Vue component. Would I need to abandon the schema if I needed props passed in, or is there another way of defining the schema, especially if I never expect it to change? Additionally, is it feasible to send data values from the two-way v-model on the main form to a custom element via the props, or should I be doing `$el('field.path')` to get these values inside the custom element instead?
2. I have a composable for making axios requests after conditionally adding an auth token into `form$.value.$vueform.services.axios`. In the larger form, I'm getting `form$` via `ref="form$"`, and then passing the ref into the composable, but is there a way of getting that same `form$` ref inside a custom element? **Nevermind, found the `form$` property! This one's solved. Above question is still relevant.**

Side note - I can't use FormSteps for this, even though theoretically it feels like the sensible route. My users would rebel if they had to click through multiple sections.

**Comment by adamberecz** [comment]
Hi Becky, thanks! I'm happy to hear that it seems to help you with such a complex case. Let's see if I can help you splitting things into different parts.

First, you mention that you're using the schema-based approach. So you mean this (or similar) by that, right?
```vue
<template>
  <Vueform :schema />
</template>

<script setup>
import { ref } from 'vue'

const schema = ref({
  a: { type: 'text' },
  b: { type: 'textarea' },
  c: { type: 'select' },
})
</script>
```

If so, have you thought about outsourcing elements into separate files, eg:

```vue
<script setup>
import { ref } from 'vue'

import a from './elements/a.js'
import b from './elements/b.js'
import c from './elements/c.js'

const schema = ref({
  a,
  b,
  c,
})
</script>
```

Could you elaborate a bit and/or give a small example of what you're looking to achieve with these?

> However, some of these custom elements would need certain values **passed in as props** as well, and you can't have props rely on other props in a Vue component. Would I need to abandon the schema if I needed props passed in, or is there another way of defining the schema, especially if I never expect it to change?

> Additionally, is it feasible to **send data values** from the two-way v-model on the main form to a custom element **via the props**, or should I be doing $el('field.path') to get these values inside the custom element instead?

**Comment by thely** [comment]
Sure - let me see if I can give a better explanation. My main form has a reactive schema. A few of my SelectElements need to be populated by either a computed element or a function, because their items depend on the values of other fields. This all works, because the schema is part of `setup`.

An example is a SelectElement whose items are a list of bookable spaces available to a selected department. The items are populated by a function, because the spaces available are based on the date selected on a DateElement. A few secondary StaticElement warnings conditionally display if some time-only DateElements are selected for times when those spaces are unavailable. I'd love to be able to split the list of spaces into a separate Vue file, because the logic for figuring out which spaces should populate and when the warnings should display is a little thorny. So: in order to split the booking logic out into a separate file, I'd imagine I need something like this, showing only the props section, not setup:

```js
props: {
    ...ObjectElement.props,
    // these three props are from the main form,
    // from three different DateElements
    date: {
      required: false,
      type: String
    },
    startTime: {
      required: false,
      type: String,
    },
    endTime: {
      required: false,
      type: String,
    },
    schema: {
      required: false,
      type: Object,
      default: () => ({
        space_select: {
          type: 'select',
          label: 'Space',
          items: (_query, input) => {
            if (props.date) {
              // this is impossible, because we can't access a prop from another prop.
              // is the only way to make this work to use el$('parent.path.date'),
              // where 'date' is on the main form?
            }
            return listOfSpaces
          }
        },
        warning_1: {
          type: 'static',
          tag: 'p',
          content: 'Can\'t use this space!',
          conditions: [
            (_form$, el$) => {
              // also based on date and time
            }
          ]
    })
    // ... etc ...
},
setup(props, context) {
  const listOfSpaces = computed(() => {
    // this can't be used to populate the 'items' above
  })
}
```

So there are two problems with splitting up the form via making custom `ObjectElement`s this way: one of them is that I can't pass props in *and* use a schema, because the schema itself is a prop, and the other being that I can't use computed elements to set anything within the schema, since `props` don't have access to anything in `setup`.

If I were in a standard Vue project, I would have done something like this:

```vue
<template>
  <ObjectElement :schema="schema" />
<template>

<script setup>
  const props = defineProps{
    // date, start, end go here
  }
  const schema = reactive({
    // essentially the schema as above
  })

  const listOfSpaces = computed(() => {
    // related computed element, which access props.date
    // to generate its list
  })
</script>
```

But I'm assuming this is outside what Vueform can do, because of how the library is structured. I imagine if I tried doing this, it would either fail outright (`ObjectElement not found` or similar), or any changes that happened in this separate Vue component wouldn't `emit`/`fire` up to the parent Vue component that contains the Vueform element.

Is the answer just that I need to lean more heavily on `el$` inside the props of the custom element for this, and redo whatever logic was relying on computed elements with basic, non-`setup` JS functions? Or is there something else I'm missing?

And as some additional information, while the list of spaces could conceivably be just one field, and would be a good candidate for making it a separate custom element, rather than a custom Group/ObjectElement, I have two other parts of the form I'd like to give this treatment:

* date/time fields (there are five in total, only three mentioned here) would be one custom element
* guest information, which is a complicated `ListElement` > `ObjectElement`, where one of the fields is a `SelectElement` that alters the items of a second `SelectElement`, and also needs the list of bookable spaces as a prop, would be another custom element
